zebra-state/src/service.rs

```diff
@@ -803,6 +803,91 @@ impl ReadStateService {
+    #[tracing::instrument(level = "debug", skip_all)]
+    fn check_best_chain_contextual_validity(
+        &self,
+        prepared: &PreparedBlock,
+    ) -> Result<(), crate::ValidateContextError> {
+        let latest_non_finalized_state = self.latest_non_finalized_state();
+
+        let Some(best_chain) = latest_non_finalized_state.best_chain() else {
+            Err(crate::ValidateContextError::NotReadyToBeValidated)?
+        };
+
+        if best_chain.non_finalized_tip_hash() != prepared.block.header.previous_block_hash {
+            Err(crate::ValidateContextError::NotReadyToBeValidated)?
+        }
+
+        check::initial_contextual_validity(
+            self.network,
+            &self.db,
+            &latest_non_finalized_state,
+            prepared,
+        )?;
+
+        // Reads from disk
+        let sprout_final_treestates =
+            check::anchors::block_fetch_sprout_final_treestates(&self.db, best_chain, prepared);
+
+        let spent_utxos = check::utxo::transparent_spend(
+            prepared,
+            &best_chain.unspent_utxos(),
+            &best_chain.spent_utxos,
+            &self.db,
+        )?;
+
+        check::anchors::block_sapling_orchard_anchors_refer_to_final_treestates(
+            &self.db, best_chain, prepared,
+        )?;
+
+        let contextual = crate::request::ContextuallyValidBlock::with_block_and_spent_utxos(
+            prepared.clone(),
+            spent_utxos.clone(),
+        )
+        .map_err(|value_balance_error| {
+            crate::ValidateContextError::CalculateBlockChainValueChange {
+                value_balance_error,
+                height: prepared.height,
+                block_hash: prepared.hash,
+                transaction_count: prepared.block.transactions.len(),
+                spent_utxo_count: spent_utxos.len(),
+            }
+        })?;
+        let mut block_commitment_result = None;
+        let mut sprout_anchor_result = None;
+
+        rayon::in_place_scope_fifo(|scope| {
+            // Clone function arguments for different threads
+            let block = Arc::clone(&contextual.block);
+
+            scope.spawn_fifo(|_scope| {
+                block_commitment_result =
+                    Some(check::anchors::block_sprout_anchors_refer_to_treestates(
+                        sprout_final_treestates,
+                        block,
+                        contextual.transaction_hashes,
+                        contextual.height,
+                    ));
+            });
+
+            scope.spawn_fifo(|_scope| {
+                sprout_anchor_result = Some(check::block_commitment_is_valid_for_chain_history(
+                    contextual.block,
+                    self.network,
+                    &best_chain.history_tree,
+                ));
+            });
+        });
+
+        block_commitment_result.expect("scope has finished")?;
+        sprout_anchor_result.expect("scope has finished")?;
+
```

Comment:
We're missing all the consensus checks that are done by this code:
https://github.com/ZcashFoundation/zebra/blob/80a6d3cdab2a9dc68e27ecfa729ba864394becb4/zebra-state/src/service/non_finalized_state.rs#L306-L308

For example, we need to apply the block chain value change to the chain value pools, and make sure none of the value pools are negative.

Reply:
Fixed in [`8420b75`](https://github.com/ZcashFoundation/zebra/pull/5870/commits/8420b75013c1a05ea11a9799d1f87bd338c3c27c).

Reply:
Thanks for fixing this up, it's clear that we're doing all the required checks now.